### PR TITLE
SAK-29546 Log presence totals per site and user

### PR DESCRIPTION
**Pull request by adrianfish**

### CODE REVIEW

**Comment by adrianfish** [comment]
I couldn't get the db conversion script change to cleanly apply, so I'll add that in another ticket.

**Comment by ern** [comment]
Wouldn't it make more sense to put this statistical data in elasticsearch vs a table?

**Comment by adrianfish** [comment]
Wouldn't that be a weird use of a search engine? It's not like anything's actually been indexed to get the totals, they're just a logical extension of the presence logging that sitestats currently, if enabled, does.

**Comment by adrianfish** [comment]
I'm also going to be writing an api call in sitestats to pull those data, so that roster can show them alongside user profiles, so it all seems to make to sense to me to be in sitestats.

**Comment by ern** [comment]
So putting the totals would not be ideal but I am more thinking of the idea of events and that information being searchable. The you could just query ES for this kind of data. Also you could save the query in ES this is called percolating see https://www.elastic.co/guide/en/elasticsearch/reference/current/search-percolate.html

**Comment by buckett** [comment]
While I think using ES for statistics is interesting I think we should concentrate on having a consistent tool and fragmenting the data store for sitestats. If we want to switch to ES for statistics I think we should look todo it on a larger scale in sitestats than just one small new feature.

**Comment by adrianfish** [comment]
It does sound interesting. Why don't you float it on the list, Earle, so there's a record? Things tend to get a bit lost in here, methinks.

File: sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/StatsUpdateManagerImpl.java [hunk 1]
```diff
@@ -1616,8 +1661,8 @@ public String toString() {
 			buff.append("firstPresEndDate: ");
 			buff.append(firstPresEndDate);
 			buff.append(" | sitePresence => ");
-			if(sitePresence != null) buff.append(sitePresence.toString());
```
Comment (buckett) [inline sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/StatsUpdateManagerImpl.java]: Just indentation change

File: sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/StatsUpdateManagerImpl.java [hunk 2]
```diff
@@ -1351,6 +1365,37 @@ private SitePresence doGetSitePresence(Session session, String siteId, String us
 		}
 		return eDb;
 	}
+
```
Comment (buckett) [inline sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/StatsUpdateManagerImpl.java]: Indentation here looks wrong.

File: sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/StatsUpdateManagerImpl.java [hunk 3]
```diff
@@ -1317,7 +1319,19 @@ private void doUpdateSitePresencesObjects(Session session, Collection<SitePresen
 			}
 		}
 	}
-	
+
+    private void doUpdateSitePresenceTotal(Session session, SitePresence sp) throws Exception {
```
Comment (buckett) [inline sitestats/sitestats-impl/src/java/org/sakaiproject/sitestats/impl/StatsUpdateManagerImpl.java]: Indentation here looks wrong.

**Comment by adrianfish** [comment]
I'll tidy up and merge, then. Thanks for reviewing.

**Comment by ern** [comment]
Hi Adrian,

Yeah I have been thinking about how under utilized ES is and how some content just sits in tables and I've been thinking of ways to change that.
I've got some ideas that I will put together and take this down a different path!
Sorry for discussing this in this issue!

**Comment by adrianfish** [comment]
@ern No worries. You are obviously at the codeface with the ES stuff :)

How persistent is the data held in ES? One of the things they want with the sitestats data is long term persistence, perhaps with warehousing. I read some of that percolator doc and kind of understood some of it. I'm doing loads of things at once at the mo, though, so I'd need to take a bit more time to study it. Thinking outside the box regarding event data is a good thing.

**Pull request merged**